### PR TITLE
NW6 | Rabia Avci | JS3 Module | [TECH ED] TV Show Project - Part 2/4 | Week 2

### PR DESCRIPTION
**Pull request by RbAvci**
## Learners, PR Template

Self checklist

- [ ] I have committed my files one by one, on purpose, and for a reason
- [ ] I have titled my PR with COHORT_NAME | FIRST_NAME LAST_NAME | REPO_NAME | WEEK 
- [ ] I have tested my changes
- [ ] My changes follow the [style guide](https://syllabus.codeyourfuture.io/guides/code-style-guide/)
- [ ] My changes meet the [requirements](./README.md) of this task

## Changelist

Search and Episode Selector added. Search is case-insensitive and the display updates immediately after each keystroke changes. It also displays how many episodes match the current search. When the user makes a selection it is taken directly to that episode in the list and you can list all episodes with a button named "show all episodes"

## Questions

Ask any questions you have for your reviewer.

### CODE REVIEW

File: script.js [hunk 1]
```diff
@@ -1,21 +1,18 @@
...
-    card.querySelector("#episode-summary").innerHTML = episode.summary;
-    rootElem.append(card);
-  }
+function createEpisodeCard(episode) {
+  const card = document.getElementById("episode-card").content.cloneNode(true);
+  const title = card.querySelector("#episode-title");
+  title.textContent = episode.name;
+  title.setAttribute("href", episode.url);
+  card.querySelector("#episode-code").textContent = episodeCode(episode.season, episode.number);
+  card.querySelector("#episode-img").src = episode.image.medium;
```
Comment (AltomHussain) [inline script.js:13]: What will happen if the image source is not available on the episode, can you think of how to hadle that ?

File: script.js [hunk 2]
```diff
@@ -24,4 +21,80 @@ function episodeCode(season, number) {
   let code = `S${s}E${num}`;
```
Comment (AltomHussain) [inline script.js:21]: Where is `s` and `num` are coming from line 24?

File: script.js [hunk 3]
```diff
@@ -24,4 +21,80 @@ function episodeCode(season, number) {
...
   return code;
 }
-window.onload = setup;
+
+const searchBox = document.getElementById("search");
+searchBox.addEventListener("input", handleSearchInput);
+function handleSearchInput(event) {
+  state.searchTerm = event.target.value;
```
Comment (AltomHussain) [inline script.js:29]: What will happen if user input is case-sensitive or case-insensitive? For example will I still get a result if I write like following words:
`Winter, WINTER, winter`?

Comment (AltomHussain) [reply script.js:29]: Remember to remove the console.log

File: script.js [hunk 4]
```diff
@@ -24,4 +21,80 @@ function episodeCode(season, number) {
...
+  console.log(state.searchTerm);
+  renderBySearch();
+}
+
+const episodeSelector = document.getElementById("episode-selector");
+episodeSelector.addEventListener("change", handleSelect);
+
+function handleSelect(event) {
+  state.selectedEpisodeId = event.target.value;
+  console.log(state.selectedEpisodeId);
```
Comment (AltomHussain) [inline script.js:39]: No need for the log.

File: script.js [hunk 5]
```diff
@@ -24,4 +21,80 @@ function episodeCode(season, number) {
   let code = `S${s}E${num}`;
   return code;
 }
-window.onload = setup;
+
+const searchBox = document.getElementById("search");
+searchBox.addEventListener("input", handleSearchInput);
```
Comment (Ara225) [inline script.js:26]: Couple of bits of code adding event listeners outside functions, I would probaly put these at the bottom of the file for tidiness  but not exactly a but issue

File: script.js (hunk 3 above)
Comment (Ara225) [inline script.js:29]: I like the use of the state hash 👍

File: script.js [hunk 6]
```diff
@@ -24,4 +21,80 @@ function episodeCode(season, number) {
...
+}
+
+function renderBySelect() {
+  searchBox.value = "";
+  state.searchTerm = "";
+  renderByFilter(filterBySelect);
+}
+
+function filterBySelect(episode) {
+  return state.selectedEpisodeId == episode.id;
```
Comment (Ara225) [inline script.js]: I would probaly inline this function - that is make it an arrow function and pass it to renderByFilter. It's only used in the one function above so ther's no need for it to be a proper funtion, but it's a matter of prefrence

File: script.js [hunk 7]
```diff
@@ -24,4 +21,80 @@ function episodeCode(season, number) {
...
+  renderByFilter(filterBySearch);
+}
+
+function filterBySearch(episode) {
+  const lowercaseName = episode.name.toLowerCase();
+  const lowercaseSummary = episode.summary.toLowerCase();
+  return lowercaseName.includes(state.searchTerm.toLowerCase()) || lowercaseSummary.includes(state.searchTerm.toLowerCase());
+}
+
+function renderByFilter(filterFunction) {
```
Comment (Ara225) [inline script.js:78]: I like how you're passing functions as arguements here - it's a really powerful tool

File: script.js [hunk 8]
```diff
@@ -24,4 +21,80 @@ function episodeCode(season, number) {
...
+  const rootElem = document.getElementById("root");
+  rootElem.innerHTML = "";
+
+  const filteredEpisodes = state.episodes.filter(filterFunction);
+  const episodeCards = filteredEpisodes.map(createEpisodeCard);
+  rootElem.append(...episodeCards);
+
+  document.getElementById("filter-info").textContent = `Displaying ${filteredEpisodes.length}/${state.episodes.length} episodes`;
```
Comment (Ara225) [inline script.js:86]: Nice touch 👍

**Review by Ara225 — COMMENTED** [review]
I like what you've done here. I might have refactored it a bit more heavily and definately added some comments. I've commented on a bit of this stuff. Overall, however, I can see you're getting really comfortable with JavaScript and starting to use some more advanced features like passing functions as arguements to funtion

**Review by AltomHussain — COMMENTED** [review]
Great work, but you did not follow the layout images provide in the project in the `example-screenshots ` folder
Your layout should look like this not the 
image

Not like how yours look
image

File: script.js (hunk 6 above)
Comment (RbAvci) [reply script.js]: Inlined the function 👍

**Review by areebsattar — APPROVED** [review]

**Pull request merged**